src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckJavaAgent.kt

```diff
@@ -0,0 +1,391 @@
+                    it.name in instrumentedClassesInTheModelCheckingMode
+                } else {
+                    true
+                }
+            }.mapNotNull { clazz ->
+                // For each class, get its original bytecode.
+                val bytes = nonTransformedClasses[clazz.name]
+                bytes?.let { ClassDefinition(clazz, it) }
+            }
+        // Redefine the instrumented classes back to their original state
+        // using the original bytecodes collected previously.
+        instrumentation.redefineClasses(*classDefinitions.toTypedArray())
+        // Clear the set of classes instrumented in the model checking mode.
+        instrumentedClassesInTheModelCheckingMode.clear()
+    }
+
+    /**
+     * Ensures that the specified class and all its superclasses are transformed.
+     *
+     * @param className The name of the class to be transformed.
+     */
+    fun ensureClassAndAllSuperClassesAreTransformed(className: String) {
+        if (INSTRUMENT_ALL_CLASSES_IN_MODEL_CHECKING_MODE) {
+            Class.forName(className)
+            return
+        }
+        if (className in instrumentedClassesInTheModelCheckingMode) return // already instrumented
+        ensureClassAndAllSuperClassesAreTransformed(Class.forName(className), Collections.newSetFromMap(IdentityHashMap()))
+    }
+
+
+    /**
+     * Ensures that the given object and all its referenced objects are transformed for Lincheck analysis.
+     * If the INSTRUMENT_ALL_CLASSES_IN_MODEL_CHECKING_MODE flag is set to true, no transformation is performed.
+     *
+     * @param testInstance the object to be transformed
+     */
+    fun ensureObjectIsTransformed(testInstance: Any) {
+        if (INSTRUMENT_ALL_CLASSES_IN_MODEL_CHECKING_MODE) {
+            return
+        }
+        ensureObjectIsTransformedImpl(testInstance, Collections.newSetFromMap(IdentityHashMap()))
+    }
+
+    /**
+     * Ensures that the given class and all its superclasses are transformed if necessary.
+     *
+     * @param clazz the class to transform
+     */
+    private fun ensureClassAndAllSuperClassesAreTransformed(clazz: Class<*>) {
+        if (INSTRUMENT_ALL_CLASSES_IN_MODEL_CHECKING_MODE) {
+            return
+        }
+        if (clazz.name in instrumentedClassesInTheModelCheckingMode) return // already instrumented
+        ensureClassAndAllSuperClassesAreTransformed(clazz, Collections.newSetFromMap(IdentityHashMap()))
+    }
+
+    /**
+     * Ensures that the given object and all its referenced objects are transformed according to the provided rules.
+     * The transformation is performed recursively, starting from the given object.
+     *
+     * @param obj The object to be ensured for transformation.
+     * @param processedObjects A set of processed objects to avoid infinite recursion.
+     */
+    private fun ensureObjectIsTransformedImpl(obj: Any, processedObjects: MutableSet<Any>) {
```

Comment:
Can remove `Impl` suffix:

```suggestion
    private fun ensureObjectIsTransformed(obj: Any, processedObjects: MutableSet<Any>) {
```